### ISSUE TITLE
reflect: decouple `Array` from `List`

### ISSUE THREAD

**Issue by soqb** [bevyengine/bevy#7121]
## What problem does this solve or what need does it fill?

Successor to #7059.
TLDR; lists should not be arrays, the nomenclature is confusing.

Additionally, the supertrait relationship is awkward and the only example within the current reflection typing. Separating them keeps our reflection trait tree flatter and therefore simpler to understand. The `clone_dynamic` collision also adds to the confusion.

## What solution would you like?

Remove the `Array` supertrait on `List`. Redeclare the methods on `Array` (`get`, `iter`, `len` etc.) onto `List`.

## What alternative(s) have you considered?

Rename the `Array` trait as proposed in #7059.

## Additional context

[Idea proposed by @MrGVSV](https://github.com/bevyengine/bevy/issues/7059#issuecomment-1374390024)

Rename advised against in #7110.

**Comment by MrGVSV** [comment]
Could you link to, reference, or copy/paste some of the additional details I mentioned in [this](https://github.com/bevyengine/bevy/issues/7059#issuecomment-1374390024) comment? I think it includes some important reasoning for making this kind of change.
